alf/utils/value_ops.py

```diff
@@ -242,32 +231,24 @@ def generalized_advantage_estimation(rewards,
         values = values.transpose(0, 1)
         step_types = step_types.transpose(0, 1)
 
-    assert values.shape[0] >= 2, "The sequence length needs to be \
-                                at least 2. Got {s}".format(s=values.shape[0])
+    assert values.shape[0] >= 2, ("The sequence length needs to be "
+                                  "at least 2. Got {s}".format(
+                                      s=values.shape[0]))
 
-    rewards = rewards[1:]
-    next_values = values[1:]
-    final_value = values[-1]
-    values = values[:-1]
-    discounts = discounts[1:]
-    step_types = step_types[:-1]
     is_lasts = (step_types == StepType.LAST).to(dtype=torch.float32)
-
-    delta = rewards + discounts * next_values - values
     weighted_discounts = discounts * td_lambda
 
-    def weighted_cumulative_td_fn(accumulated_td, weighted_discount, td,
-                                  is_last):
-        return (1 - is_last) * (td + weighted_discount * accumulated_td)
+    advs = torch.zeros(rewards.shape, dtype=rewards.dtype)
 
-    advantages = torch.zeros(rewards.shape, dtype=rewards.dtype)
-    adv = 0
-    for t in reversed(range(rewards.shape[0])):
-        adv = weighted_cumulative_td_fn(adv, weighted_discounts[t], delta[t],
-                                        is_lasts[t])
-        advantages[t] = adv
+    with torch.no_grad():
+        for t in reversed(range(rewards.shape[0] - 1)):
+            delta = rewards[t + 1] + \
```

Comment:
Do the following before the loop to utilize vector computation as more as possible.
```
delta = reward[1:] + discounts[1:] * values[1:] - values[:-1] 
```

Reply:
Changed to vectorized computation for delta.